Hidden powers of Bitcoin Blog

sandipndev opened the pull request:
In this blog, I have introduced the idea of a descriptor, made a multisig descriptor wallet and tested miniscript policies.
This is a very beginner's take on all these topics.
Hope you guys like it.

sandipndev committed:
added hidden powers of bitcoin blog

notmandatory commented:
I did a quick read and looks good so far, I'll do a more thorough read tomorrow. In the meantime please add an authors file for yourself in: `content/authors/<your-name>/_index.md`

sandipndev committed:
added author details

rajarshimaitra commented on content/authors/sandipndev/_index.md:
```diff
@@ -0,0 +1,8 @@
+---
+name: Sandipan Dey
```
File name should be `sandipan-dey`. Otherwise hugo wont link the author to the page.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+### Script
+
+Bitcoin supports [Script](https://en.bitcoin.it/wiki/Script), a **stack-based** lightweight programming language.
+Any script written in **Script** *(pun intended)* contains only `OP_*` codes that Bitcoin Full Nodes understand and process.
```
Script contains both `OP_*` codes and raw byte arrays.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+
+Bitcoin supports [Script](https://en.bitcoin.it/wiki/Script), a **stack-based** lightweight programming language.
+Any script written in **Script** *(pun intended)* contains only `OP_*` codes that Bitcoin Full Nodes understand and process.
+Currently, there are `186` op-codes in use.
```
`186` seems too much. As per my counting there are 117 opcodes (including OP_NOPs).  

Also maybe its better to link the wiki page for scripts instead the core code repo? Thats more descriptive and maybe helpful for new learners. https://en.bitcoin.it/wiki/Script

sandipndev replied:
Yes, this was a miscalculation on my part, `117` is correct 😅
And yup, the Wiki Page has better descriptions on each of the OP codes., adding it!

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+
+A transaction is considered valid only when an Output Script returns `true` at the end of execution.
+Output Scripts define the conditions under which the coins associated with them can be spent.
+Really, to spend the output of a particular coin implies finding an input *(signature)* that makes a particular script evaluate to `true`.
```
Techincally "Output script" doesn't evaluate to true on its own. Below is a re-framing suggestion of the above paragraph. 

 ```suggestion
 A transaction is considered valid only when the Script returns `true` at the end of execution.
Output Script (aka scriptpubkey) define the conditions under which coins associated with them can be spent. To spend a particular coin implies finding an input script (aka scriptsig) such that a script made out of concatenation of `scriptsig + scriptpubkey` evaluates to `true`.
```

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+1. `Pay Someone (p2pkh/p2wpkh)` - A specific public key must sign to use the coins.
+2. `Escrow (2-of-3-multisig)` - Two parties need to sign together to use the coins.
+3. `Vault (locked)` - A specific key will not be able to use the coins for some time but another master key will always be able to use them.
+4. `HTLC` - The receiver needs to ACK the tx before a time, else the coins are transferred back to the payee.
```
Suggestions inline. 
```suggestion
1. `Pay Someone (p2pkh/p2wpkh)` - A specific public key must sign to spend the coins.
2. `Escrow (2-of-3-multisig)` - Two parties need to sign together to spend the coins.
3. `Vault (locked)` - A specific key will not be able to spend the coins until a timeout but another master key will always be able to spend them.
4. `HTLC` - The receiver needs disclose a secret before a timeout, else the coins are transferred back to the payee.
```

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+##### Motivation for Policies
+
+Unfortunately, due to its low-level and unusual stack-based nature, Script is pretty hard to reason about and use.
+Despite being around since Bitcoin creation, writing and understanding Script is not trivial.
```
```suggestion
Despite being around since Bitcoin's creation, writing and understanding Script is not trivial.
```

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+Another very important goal of Miniscript is to replace any key used in a policy with another policy.
+This is important because people might have multiple keys and complicated timelocks in their existing setup.
+While signing a new policy, they would want to use this existing setup of theirs and also generate addresses for this setup.
+This is something accomplished using something called **Output Descriptors** which we will read about later.
```
This paragraph seemed a little confusing to me. Are you mentioning that we can replace a "key" or "value" inside a policy with another policy? Maybe I am wrong, but is this in general true?

In the last line you mentioned "Descriptors", so were you talking about replacing one "key" with another "key"? 

Nit:
  ```suggestion
Another very important goal of Miniscript is to replace any key used in a policy with another policy.
This is important because people might have multiple keys and complicated timelocks in their existing setup.
While signing a new policy, they would want to use their existing setup to also generate addresses for this new setup.
This is accomplished using something called **Output Descriptors** which we will get into in next section.
```

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+##### Example Policies
+
+Here are the Miniscript Policies for the examples we looked at earlier. 
+Note `A`, `B`, `C` are placeholders for keys *(or output descriptors)* involved in the tx.
```
You cant put a descriptor inside miniscript, its other way around. 

Descriptors are high level description of scriptpubkey (p2pkh, p2sh etc). And miniscripts are semantics that describes the actual redeemscript. In general you have `Descriptor(Miniscript)` format.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+
+Sections:
+1 - function type (here, P2PK)
+2 - derivation path of this key
```
Nit:
```suggestion
2 - master key fingerprint and derivation path from master
```

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
````diff
@@ -0,0 +1,508 @@
+```
+
+This was an example of a simple descriptor using one function but they can also be used in conjugation with multiple functions like `combo`, `multi` etc.
+The concept of descriptor came into existence in 2018 and since then, a lot of wallets have added support for descriptors.
````
Might be a good idea to include a link for descriptor doc here somewhere. https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+
+# create the transaction, can be started by anyone
+PSBT=$(bdk-cli -n regtest wallet --server tcp://127.0.0.1:60401 -w K2 -d $K2_DESC \
+	create_tx --to "<address-from-last-step>:5555555" | jq -r ".psbt")
```
Instead of sending address from last step its better to store it in a variable. Its done like that everywhere else.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+
+So now we have definitive proof that descriptors can make wallets portable. That single string will be able to make any wallet generate the same set of addresses and hence they can sync and broadcast transactions in the same manner!
+
+#### Making a Secure Descriptor for Funds
```
It seems this a "Multisig" transaction section. "Secure" is generic term and can mean many things.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+If an employee stays with that company for over 2 months, the employee would get 1 BTC as a reward.
+This would be a smart contract between the company and an employee.
+The employee should be able to see that he would get his funds after a year.
+The company would require confidence that the employee would not be able to withdraw the reward before 1 year has passed.
```
Shouldn't the employee get the funds after 2 months (not 1 year)?

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+
+#### Keys and Generating Addresses
+
+Let us first generate an XPRV and create the wpkh wallet descriptor
```
Maybe before this step its better to include installation instruction for bdk-cli? 

bdk-cli would also needs a backend to work with. In latest update the default electrum backend is removed. So in order to use bdk-cli for this tutorial they need to compile it with `--features electrum --features compiler`.

sandipndev replied:
I tried `cargo install bdk-cli --features electrum --features compiler` but this didn't complete successfully.
It's ending in: 
```
error: no binaries are available for install using the selected features
```

rajarshimaitra replied:
Ah that's because the new version 0.2.1 of bdk-cli hasn't been tagged yet. cc @notmandatory . Currently cargo is trying to install 0.2.0, which requires the `repl` feature and is not included by default.

This works with 0.2.0 `cargo install bdk-cli --features electrum --features compiler --features repl`.

But we should get the latest bdk-cli tagged before releasing this tutorial. Otherwise the instructions can break in future.

rajarshimaitra replied:
Also curious, what version of bdk-cli did you use in the tutorial?

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+It's not possible to block them after the lock time has passed, atleast not in a single policy.
+
+Surely, after two months, the funds can be unlocked by the employee but before that, the company can revoke the funds.
+Let us compile this policy down to miniscript.
```
The policy is compiled down to a `descriptor`. Not another miniscript.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
````diff
@@ -0,0 +1,508 @@
+
+```bash
+# The Descriptor will be on the log, the E and C are placeholders
+❯ miniscriptc "or(99@and(pk(E),older(8640)),pk(C))" sh-wsh
````
This probably old version of bdk-cli. Currently it doesn't produce `miniscriptc` anymore. we have to use `bdk-cli compile` command.

rajarshimaitra replied:
Also please confirm this with `v0.2.0` as far as I can see there is no `miniscriptc` binary anymore.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
````diff
@@ -0,0 +1,508 @@
+
+The Miniscript policy for this contract would be as follows:
+```
+or(99@and(pk(E),older(8640)),pk(C))
````
Either I am reading it wrong or 8640 is blocks worth of 1 month, not 2.

sandipndev replied:
Block Time = 10 mins
2 Months = 2*30*24*60 mins = 86400 mins
n(Blocks) = 8640

Am I doing it wrong somewhere? 😅

rajarshimaitra replied:
Ah my bad.. I made stupid mistake in my calculation. This is correct.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
````diff
@@ -0,0 +1,508 @@
+
+So the compiled miniscript is
+```
+sh(wsh(andor(pk(E),older(8640),pk(C))))
````
This is a descriptor. The miniscript is inside the descriptor.

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
```diff
@@ -0,0 +1,508 @@
+They make it easier to deal with Multisig or complicated key setups.
+Descriptors are super portable and can be easily used by any wallet to determine the list of all addresses that can be generated from the same.
+This feature creates a common stage for all Bitcoin apps and software.
+It's an abstract way to replace keys.
```
To describe a little clearly the difference between a Descriptor and Miniscript, maybe you can add something along this line: 

"A bitcoin scriptpubkey can be of various types, ex P2PKH, P2SH, P2SH-P2WPKH etc. A descriptor is a simple description of what scriptpubkey scheme to be used for a given policy. A Descriptor can include a single pubkey within itself, or an entire miniscript policy. The miniscript policy will be used to derive the resulting redeemscript (the actual executable script), where as the descriptor describes how the redeemscript will be encumbered within the scriptpubkey."

rajarshimaitra commented on content/blog/2021/hidden_powers_of_bitcoin.md:
````diff
@@ -0,0 +1,508 @@
+5 - checksum for the descriptor
+```
+
+This was an example of a simple descriptor using one function but they can also be used in conjugation with multiple functions like `combo`, `multi` etc.
````
`combo` and `multi` are two different things. `combo` is a descriptor which describes scriptpubkey of any type ex: P2PKH, P2SH etc.

`multi` is basically a miniscript which describes a m-of-n multisig script. When used as descriptor it will make the scriptpubkey as the raw multisig script. But its better to think it as a miniscript than a descriptor, because such raw multisigs are never used in practice.

In wild you will see stuffs like `wsh(multi(2, PKA, PKB, PKC))`. The `wsh` part is the descriptor describing its a P2WSH output, and `multi` part is the miniscript describing the actual spending condition.

Ya it gets messy with terminology in bitcoin scripts. But if we aren't careful, this can cause more confusion to readers.

sandipndev replied:
True, I added this section in the blog because it makes the concept clearer.

rajarshimaitra reviewed (commented):
Thanks for such a detail walk through of minsicripts with bdk-cli. Its always good to have more miniscript tutorials as its a complicated concept.

untested ACK. 

Below are some comments and suggestions.

sandipndev committed:
Update content/blog/2021/hidden_powers_of_bitcoin.md

Co-authored-by: Raj <36541669+rajarshimaitra@users.noreply.github.com>

Update content/blog/2021/hidden_powers_of_bitcoin.md

Co-authored-by: Raj <36541669+rajarshimaitra@users.noreply.github.com>

Update content/blog/2021/hidden_powers_of_bitcoin.md

Co-authored-by: Raj <36541669+rajarshimaitra@users.noreply.github.com>

Update content/blog/2021/hidden_powers_of_bitcoin.md

Co-authored-by: Raj <36541669+rajarshimaitra@users.noreply.github.com>

Update content/blog/2021/hidden_powers_of_bitcoin.md

Co-authored-by: Raj <36541669+rajarshimaitra@users.noreply.github.com>

fix to requested changes

rajarshimaitra commented:
Can you squash the commits? Will review again after that. Thanks for putting me in co-author, wasn't required.

rajarshimaitra reviewed (commented):
@sandipndev sorry it took me a long time to get back to this.

I went through the tutorial once again and the text looks good to me. Though I would like some framing change near the descriptor and miniscript explanation section.

The code section doesn't work anymore because `bitcoindevkit:electrs` docker image is removed as we are not maintaining them anymore. This needs to be changed into either local `electrs` or `bitcoind`. I think regtest `bitcoind` would be enough and easy to setup. We should have a separate tutorial for setting up regtest and link it in other tutorials for regtest setup. I think most of the tutorials will use regtest backend.

@ConorOkus This is a really valuable literature and would be a very good addition to our updated doc site. There are not enough descriptor tutorials out there, and this one really showcases the power of bdk for playing with descriptors. We can get this to finish line with little push.

The works needed are following
 - fix some explanation framing.
 - change elctrs docker to bitcoind
 - port to new doc format

I don't wanna bother @sandipndev over this as this has been hanging for quite some time now and the remaining work is mostly logistics.

I am already working on the above fixes and would either do a separate PR with @sandipndev as co-author or just push my changes to this branch, depending on which is easier for porting to new website. @ConorOkus your suggestion here would be helpful.

rajarshimaitra commented:
I have updated the blog post with latest changes to make the codes work. I seem to not have adequate permission to push to this PR.

So I guess the best thing to do here is open another PR on current master. Will do that soon.

rajarshimaitra commented:
Updated PR opened at https://github.com/bitcoindevkit/bitcoindevkit.org/pull/82

sandipndev commented:
Thanks for continuing the work :D

ConorOkus commented:
Sorry was late to this, great stuff!
